File: skills/cp-solver/references/implementations/matrix-exp/matrix_exp.py
"""
Matrix Exponentiation — Fast matrix power O(K³ log N)
Dùng cho: linear recurrences với N lớn (10^18), DP transitions

Usage:
    mat = Matrix([[1,1],[1,0]])
    result = mat ** n       # A^n mod MOD
    val = result[0][0]
"""
import sys
input = sys.stdin.readline

MOD = 10**9 + 7


class Matrix:
    def __init__(self, data):
        self.data = data
        self.n = len(data)
        self.m = len(data[0])

    def __mul__(self, other):
        n, m, p = self.n, self.m, other.m
        result = [[0] * p for _ in range(n)]
        for i in range(n):
            for k in range(m):
                if self.data[i][k] == 0: continue
                for j in range(p):
                    result[i][j] = (result[i][j] + self.data[i][k] * other.data[k][j]) % MOD
        return Matrix(result)

    def __pow__(self, n):
        """A^n mod MOD — O(K³ log n)"""
        sz = self.n
        # Identity matrix
        result = Matrix([[1 if i == j else 0 for j in range(sz)] for i in range(sz)])
        A = Matrix([row[:] for row in self.data])
        while n > 0:
            if n & 1:
                result = result * A
            A = A * A
            n >>= 1
        return result

    def __getitem__(self, idx):
        return self.data[idx]

    def __repr__(self):
        return '\n'.join(str(row) for row in self.data)


def fibonacci(n):
    """F(n) mod MOD sử dụng matrix exponentiation — O(log n)"""
    if n <= 1: return n
    M = Matrix([[1, 1], [1, 0]])
    Mn = M ** (n - 1)
    return Mn[0][0]


def linear_recurrence(coeffs, initial, n):
    """
    Tính f(n) với recurrence: f(i) = sum(coeffs[j] * f(i-1-j))
    coeffs: list length K
    initial: f(0), f(1), ..., f(K-1)
    n: giá trị cần tính

    Ví dụ Fibonacci: coeffs=[1,1], initial=[0,1], n=10 → f(10)=55
    """
    k = len(coeffs)
    if n < k: return initial[n]

    # Companion matrix
    M = [[0] * k for _ in range(k)]
    for j in range(k):
        M[0][j] = coeffs[j]
    for i in range(1, k):
        M[i][i-1] = 1

    # State vector [f(k-1), f(k-2), ..., f(0)]
    Mn = Matrix(M) ** (n - k + 1)

    result = 0
    for j in range(k):
        result = (result + Mn[0][j] * initial[k-1-j]) % MOD
    return result


# ===== USAGE EXAMPLE =====
def solve():
    n = int(input())
    print(fibonacci(n))


if __name__ == '__main__':
    solve()

File: skills/cp-solver/references/implementations/matrix-exp/matrix_exp.cpp
// Matrix Exponentiation — Fast matrix power O(K³ log N)
// Dùng cho: linear recurrences với N lớn (10^18), DP transitions
// Template: Fibonacci, linear recurrence, graph powers

#include <bits/stdc++.h>
using namespace std;

const long long MOD = 1e9 + 7;
typedef vector<vector<long long>> Matrix;

Matrix multiply(const Matrix &A, const Matrix &B) {
  int n = A.size(), m = A[0].size(), p = B[0].size();
  Matrix C(n, vector<long long>(p, 0));
  for (int i = 0; i < n; i++)
    for (int k = 0; k < m; k++)
      if (A[i][k])
        for (int j = 0; j < p; j++)
          C[i][j] = (C[i][j] + A[i][k] * B[k][j]) % MOD;
  return C;
}

// Matrix power: A^n mod MOD in O(K³ log n)
Matrix power(Matrix A, long long n) {
  int sz = A.size();
  Matrix result(sz, vector<long long>(sz, 0));
  for (int i = 0; i < sz; i++)
    result[i][i] = 1; // identity

  while (n > 0) {
    if (n & 1)
      result = multiply(result, A);
    A = multiply(A, A);
    n >>= 1;
  }
  return result;
}

// ===== EXAMPLE: Fibonacci F(n) mod MOD =====
// F(n+1)   [1 1] [F(n)  ]
// F(n)   = [1 0] [F(n-1)]
long long fibonacci(long long n) {
  if (n <= 1)
    return n;
  Matrix M = {{1, 1}, {1, 0}};
  Matrix Mn = power(M, n - 1);
  return Mn[0][0]; // F(n) = M^(n-1)[0][0] * F(1) + M^(n-1)[0][1] * F(0)
}

// ===== EXAMPLE: Count paths of length K in graph =====
// adj_matrix^K[i][j] = số đường đi độ dài K từ i đến j
Matrix count_paths(Matrix adj, long long K) { return power(adj, K); }

int main() {
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);

  long long n;
  cin >> n;
  cout << fibonacci(n) << "\n";

  return 0;
}
